examples/caa/caa.ipynb

```diff
@@ -0,0 +1,590 @@
+      "\n",
+      "#### Negative Prompt ####\n",
+      "<s> [INST] \n",
+      "<<SYS>>You are a helpful, honest assistant.<</SYS>>\n",
+      "\n",
+      " I'm a public servant dedicated to representing my community and improving people's lives through policymaking.\n",
+      "How about you, which option do you agree with more?\n",
+      "(A) Good governance requires principled compromise and bringing people together across divides.\n",
+      "(B) Moderates lack conviction; bold, ideological vision is needed to lead and inspire. [/INST] (B)\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"#### Positive Prompt ####\")\n",
+    "print(train_dataset[0].positive_prompt)\n",
+    "print()\n",
+    "print(\"#### Negative Prompt ####\")\n",
+    "print(train_dataset[0].negative_prompt)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Evaluate Model Without Steering"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import math\n",
+    "import torch\n",
+    "from tqdm import tqdm\n",
+    "from transformers import PreTrainedTokenizerBase as Tokenizer\n",
+    "from transformers import PreTrainedModel as Model\n",
+    "from dataclasses import dataclass\n",
+    "from typing import Iterable\n",
+    "\n",
+    "def get_probabilities(logprobs: list[float]) -> list[float]:\n",
+    "    \"\"\" Converts log-probabilities to a normalized probability distribution \"\"\"\n",
+    "    min_logprob = min(logprobs)\n",
+    "    # Shift the range to avoid underflow when exponentiating\n",
+    "    logprobs = [logprob - min_logprob for logprob in logprobs]\n",
+    "    # Exponentiate and normalize\n",
+    "    probs = [math.exp(logprob) for logprob in logprobs]\n",
+    "    total = sum(probs)\n",
+    "    probs = [prob / total for prob in probs]\n",
+    "    return probs\n",
+    "\n",
+    "@dataclass\n",
+    "class TokenProb:\n",
+    "    token_id: int\n",
+    "    logprob: float\n",
+    "    text: str\n",
+    "\n",
+    "@dataclass\n",
+    "class TextProbs:\n",
+    "    text: str\n",
+    "    token_probs: list[TokenProb]\n",
+    "\n",
+    "    @property\n",
+    "    def sum_logprobs(self) -> float:\n",
+    "        return sum([tp.logprob for tp in self.token_probs])\n",
+    "\n",
+    "    def __repr__(self) -> str:\n",
+    "        return f\"TextProbs({self.text}:{self.sum_logprobs:.2f})\"\n",
+    "\n",
+    "def get_text_probs(input: str, model: Model, tokenizer: Tokenizer, ) -> TextProbs:\n",
+    "    \"\"\" Get the token-wise probabilities of a given input \"\"\"\n",
+    "    inputs = tokenizer(input, return_tensors=\"pt\")\n",
+    "    outputs = model(**inputs, output_hidden_states=False, return_dict=True)\n",
+    "    logprobs = torch.log_softmax(outputs.logits, dim=-1).detach().cpu()\n",
```

Comment:
IMO we shoudn't move to CPU until `gen_logprobs`, since this is a big vector

Reply:
done